diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt

```diff
@@ -0,0 +1,160 @@
+ */
+class CollapseIfStatementsRule(configRules: List<RulesConfig>) : DiktatRule(
+    "collapse-if",
+    configRules,
+    listOf(
+        Warnings.COLLAPSE_IF_STATEMENTS
+    )
+) {
+    private val configuration by lazy {
+        CollapseIfStatementsConfiguration(
+            configRules.getRuleConfig(Warnings.COLLAPSE_IF_STATEMENTS)?.configuration ?: emptyMap()
+        )
+    }
+
+    // We hold the warnings, which we raised, since in case of multi nested if-statement,
+    // there are could be several identical warning for one line
+    private val listOfWarnings: MutableSet<placeOfwarningForCurrentNode> = mutableSetOf()
+
+    override fun logic(node: ASTNode) {
+        if (node.elementType == IF) {
+            process(node)
+        }
+    }
+
+    private fun process(node: ASTNode) {
+        val startCollapseFromLevel = configuration.startCollapseFromNestedLevel
+        val listOfNestedNodes: Stack<ASTNode> = Stack()
+
+        var nestedIfNode = findNestedIf(node)
+        while (nestedIfNode != null) {
+            if (nestedIfNode !in listOfNestedNodes) {
+                listOfNestedNodes.push(nestedIfNode)
+            }
+            nestedIfNode = findNestedIf(nestedIfNode)
+        }
+        val nestedLevel = listOfNestedNodes.size + 1
+        if (nestedLevel < startCollapseFromLevel) {
+            return
+        }
+        while (listOfNestedNodes.isNotEmpty()) {
+            val currNode = listOfNestedNodes.pop()
+            // Since the external `if` statement is not the direct parent,
+            // we need multiple steps to take the required one
+            // BLOCK -> THEN -> IF
+            val currParentNode = currNode.treeParent.treeParent.treeParent
+            if (listOfWarnings.add(currNode.startOffset to currNode)) {
+                Warnings.COLLAPSE_IF_STATEMENTS.warnAndFix(
+                    configRules, emitWarn, isFixMode,
+                    "avoid using redundant nested if-statements", currNode.startOffset, currNode
+                ) {
+                    collapse(currParentNode, currNode)
+                }
+            }
+        }
+    }
+
+    private fun findNestedIf(parentNode: ASTNode): ASTNode? {
+        val parentThenNode = (parentNode.psi as KtIfExpression).then?.node
+        val nestedIfNode = parentThenNode?.findChildByType(IF) ?: return null
+        // We won't collapse statements, if nested `if` statement have `else` node
+        (nestedIfNode.psi as KtIfExpression).`else`?.node?.let {
+            return null
+        }
+        // We monitor which types of nodes are followed before nested `if`
+        // and we allow only a limited number of types to pass through.
+        // Otherwise discovered `if` it is not nested
+        val listOfNodesBeforeNestedIf = parentThenNode.getChildren(null).takeWhile { it.elementType != IF }
+        val allowedTypes = listOf(LBRACE, WHITE_SPACE, RBRACE, KDOC, BLOCK_COMMENT, EOL_COMMENT)
```

Comment:
Which node types can we discover here? I thought, in `IF` node there can't be any non-comment nodes before `if`, can they?

Reply:
Here we consider the THEN block, and in this case externa IF is not the direct parent for nested IF.
At least first child will be: WHITE_SPACE and braces. However, it could be comments too, but this case will be considered in https://github.com/cqfn/diKTat/issues/817